sql/hive/src/test/scala/org/apache/spark/sql/sources/HadoopFsRelationTest.scala

```diff
@@ -126,61 +126,60 @@ abstract class HadoopFsRelationTest extends QueryTest with SQLTestUtils with Tes
     } else {
       Seq(false)
     }
-    // TODO: Support new parser too, see SPARK-26374.
-    withSQLConf(SQLConf.LEGACY_TIME_PARSER_ENABLED.key -> "true") {
-      for (dataType <- supportedDataTypes) {
-        for (parquetDictionaryEncodingEnabled <- parquetDictionaryEncodingEnabledConfs) {
-          val extraMessage = if (isParquetDataSource) {
-            s" with parquet.enable.dictionary = $parquetDictionaryEncodingEnabled"
-          } else {
-            ""
-          }
-          logInfo(s"Testing $dataType data type$extraMessage")
-
-          val extraOptions = Map[String, String](
-            "parquet.enable.dictionary" -> parquetDictionaryEncodingEnabled.toString
-          )
-
-          withTempPath { file =>
-            val path = file.getCanonicalPath
-
-            val seed = System.nanoTime()
-            withClue(s"Random data generated with the seed: ${seed}") {
-              val dataGenerator = RandomDataGenerator.forType(
-                dataType = dataType,
-                nullable = true,
-                new Random(seed)
-              ).getOrElse {
-                fail(s"Failed to create data generator for schema $dataType")
-              }
-
-              // Create a DF for the schema with random data. The index field is used to sort the
-              // DataFrame.  This is a workaround for SPARK-10591.
-              val schema = new StructType()
-                .add("index", IntegerType, nullable = false)
-                .add("col", dataType, nullable = true)
-              val rdd =
-                spark.sparkContext.parallelize((1 to 10).map(i => Row(i, dataGenerator())))
-              val df = spark.createDataFrame(rdd, schema).orderBy("index").coalesce(1)
-
-              df.write
-                .mode("overwrite")
-                .format(dataSourceName)
-                .option("dataSchema", df.schema.json)
-                .options(extraOptions)
-                .save(path)
-
-              val loadedDF = spark
-                .read
-                .format(dataSourceName)
-                .option("dataSchema", df.schema.json)
-                .schema(df.schema)
-                .options(extraOptions)
-                .load(path)
-                .orderBy("index")
-
-              checkAnswer(loadedDF, df)
+
+    for (dataType <- supportedDataTypes) {
+      for (parquetDictionaryEncodingEnabled <- parquetDictionaryEncodingEnabledConfs) {
+        val extraMessage = if (isParquetDataSource) {
+          s" with parquet.enable.dictionary = $parquetDictionaryEncodingEnabled"
+        } else {
+          ""
+        }
+        logInfo(s"Testing $dataType data type$extraMessage")
+
+        val extraOptions = Map[String, String](
+          "parquet.enable.dictionary" -> parquetDictionaryEncodingEnabled.toString,
+          "timestampFormat" -> "yyyy-MM-dd'T'HH:mm:ss.SSSXXXXX"
```

Comment:
Ah, this is the fix? Great, looks good. Should this... be a default? You know much more about it than I do right now.

Reply:
> Should this... be a default? 

I am not completely sure that `XXXXX` should be default pattern for time zone offset because:
1. It could potentially break existing user apps. For example `SimpleDateFormat` cannot recognize such zone offset.
2. Legacy parser is still supported, and `SimpleDateFormat` will throw `IllegalArgumentException` if the number of pattern letters is 4 or more.
3. I think it is rare case when old timestamps (< 1582 year) are written with time zone offset, and should be read back.

Reply:
I am not against of the pattern `XXXXX` by default in CSV/JSON datasources. I would just try to enabled it after the legacy parsed is removed.

Reply:
OK, I suppose all these changes are going to go in together. I'm still kinda curious where the parsing logic is and how it could use this format... I dug around the code for a while but got lost finding where it is.

Reply:
> what code can use this pattern to parse -- do you know?
> I'm still kinda curious where the parsing logic is and how it could use this format...

I guess it is inside of `DateTimeFormatterBuilder.java` in the method `parse`:
```java
            char sign = text.charAt(position);  // IOOBE if invalid position
            if (sign == '+' || sign == '-') {
                // starts
                int negative = (sign == '-' ? -1 : 1);
                int[] array = new int[4];
                array[0] = position + 1;
                if ((parseNumber(array, 1, text, true) ||
                        parseNumber(array, 2, text, type >=3) ||
                        parseNumber(array, 3, text, false)) == false) {
                    // success
                    long offsetSecs = negative * (array[1] * 3600L + array[2] * 60L + array[3]);
                    return context.setParsedField(OFFSET_SECONDS, offsetSecs, position, array[0]);
                }
            }
```

Reply:
The method `parsePattern` in the same file parse the pattern including zone offset:
```java
} else if (cur == 'X') {
  if (count > 5) {
   throw new IllegalArgumentException("Too many pattern letters: " + cur);
 }
   appendOffset(OffsetIdPrinterParser.PATTERNS[count + (count == 1 ? 0 : 1)], "Z");
 }
```

Reply:
or do you mean parsing logic inside of JSON datasource?

Reply:
OK so I guess we have some custom code to handle the (slightly nonstandard, but important) timezone with seconds offset. I mean the tests pass, so it's doing it right somewhere, just kinda wanted to understand why it would work at all to make sure I'm not missing something -- because as you say it can't be using the pattern you added, it's not valid for parsing. This sounds OK to me.